aioxmpp/ibr/service.py

```diff
@@ -0,0 +1,191 @@
+########################################################################
+# File name: service.py
+# This file is part of: aioxmpp
+#
+# LICENSE
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+#
+########################################################################
+
+import asyncio
+import aioxmpp
+import logging
+from aioxmpp.service import Service
+from . import xso
+
+
+logger = logging.getLogger(__name__)
+
+
+@asyncio.coroutine
+def get_registration_fields(xmlstream,
+                            timeout=60,
+                            ):
+    """
+    A query is sent to the server to obtain the fields that need to be
+    filled to register with the server.
+
+    :param xmlstream: Specifies the stream connected to the server where
+                      the account will be created.
+    :type xmlstream: :class:`aioxmpp.protocol.XMLStream`
+
+    :param timeout: Maximum time in seconds to wait for an IQ response, or
+                    :data:`None` to disable the timeout.
+    :type timeout: :class:`~numbers.Real` or :data:`None`
+
+    :return: :attr:`list`
+    """
+
+    iq = aioxmpp.IQ(
+        to=aioxmpp.JID.fromstr(xmlstream._to),
+        type_=aioxmpp.IQType.GET,
+        payload=xso.Query()
+    )
+    iq.autoset_id()
+
+    reply = yield from aioxmpp.protocol.send_and_wait_for(xmlstream,
+                                                          [iq],
+                                                          [aioxmpp.IQ],
+                                                          timeout=timeout)
+    return reply.payload
+
+
+@asyncio.coroutine
+def register(xmlstream,
+             query_xso,
+             timeout=60,
+             ):
+    """
+    Create a new account on the server.
+
+    :param query_xso: XSO with the information needed for the registration.
+    :type query_xso: :class:`xso.Query`
+
+    :param xmlstream: Specifies the stream connected to the server where
+                      the account will be created.
+    :type xmlstream: :class:`aioxmpp.protocol.XMLStream`
+
+    :param timeout: Maximum time in seconds to wait for an IQ response, or
+                    :data:`None` to disable the timeout.
+    :type timeout: :class:`~numbers.Real` or :data:`None`
+    """
+    iq = aioxmpp.IQ(
+        to=aioxmpp.JID.fromstr(xmlstream._to),
```

Comment:
I’m not 100% happy with the use of the private attribute here. But fixing that is out-of-scope for this PR, so we’ll leave it at this.